### DISCUSSION TITLE
Select Custom Table Column save datas

### DISCUSSION THREAD

**Discussion by antoinemy** [filamentphp/filament#3839] [Help] [answered]
Hi,

I'm new to Filament and Livewire, I need some help please.

I created a custom column with a select that changes the state of the relationship in live and why not a validation by modal maybe later :

**/resources/views/tables/select-switcher-column.blade.php**

```
@php
    $state = $getState();
@endphp

<select>
    @foreach ($getOptions() as $id => $label)
        <option
            value="{{ $id }}"
            @if($id === $state->id) selected="selected" @endif
        >
            {{ $label }}
        </option>
    @endforeach
</select>
```

**/app/Tables/Columns/SelectSwitcherColumn.php**

I created a function **getOptions()** which returns the options of the select from the initialization of the table so that we can reuse the select on several pages. I took over how you did for the forms...

```
protected array | Arrayable | string | Closure $options = [];
protected bool | Closure | null $isOptionDisabled = null;

public function options(array | Arrayable | string | Closure $options): static
{
    $this->options = $options;

    return $this;
}

public function getOptions(): array
{
    $options = $this->evaluate($this->options);
    
    if (is_string($options) && function_exists('enum_exists') && enum_exists($options)) {
        $options = collect($options::cases())->mapWithKeys(static fn ($case) => [($case?->value ?? $case->name) => $case->name]);
    }
    
    if ($options instanceof Arrayable) {
        $options = $options->toArray();
    }
    
    return $options;
}
```

**/app/Filament/DeviceResource.php**

So I call that from the table

```
SelectSwitcherColumn::make('status')
    ->label('Status')
    ->options(StatusDevice::all()->pluck('name', 'id')->toArray())
```

So far everything is working and this is where I am stuck with passing data with Livewire. Can someone give me the best practice to do because I managed to do something like this that works:

```
<select wire:change="$emit('statusChange', {{ $getId() }}, $event.target.value)">
```

But that's DIY and I figure there must be a best practice with a Livewire file. But I don't understand how Livewire works with Filament in a Resource, I find the documentation not very clear for beginners...

Thank's, Antoine

**Comment by danharrin** [comment, marked as answer]
So in this case, the List page is the Livewire component, so you'd just need a trait that you add to all list pages.

The trait should contain a public method that you can call to update the value of the field. Maybe instead of emitting an event, try:

`wire:change="updateColumnValue('{{ $getRecord()->getKey() }}', '{{ $getName() }}', $event.target.value)"`

And then in the trait:

```php
public method updateColumnValue($record, $column, $value)
{
    $this->resolveTableRecord($record)?->update([$column => $value]);
}
```

This behaviour will be built into Filament soon, its on our roadmap

**Reply by antoinemy** [reply]
Thanks, it works fine!

I would like to know how I can make the **->searchable()** and **->sortable()** methods work correctly when I pass a relation in the **::make()** method of my custom column? After reading the documentation, I should simply pass them the attribute I want as a parameter;

In my case: ['status_device.name'] but it gives me an error, yet my relationship seems correct?

**App/Filament/Resources/DeviceResource.php**

```
SelectSwitcherColumn::make('status_device')
    ->label('Etat')
    ->options(StatusDevice::all()->pluck('name', 'id')->toArray())
    ->sortable(['status_device.name'])
    ->searchable(['status_device.name')
    ->toggleable();
})
```

**App/Models/Device.php**

```
protected $fillable = [
    'status_device_id',
    'name',
];

public function status_device()
{
    return $this->belongsTo(StatusDevice::class, 'status_device_id');
}
```

**App/Models/StatusDevice.php**

```
protected $fillable = [
    'name',
];
```

**Reply by danharrin** [reply]
Filament only knows if it's a relationship column if the column name format is `relationship.column`, so try changing it to `SelectSwitcherColumn::make('status_device.name')`. Then you wont need the custom sortable and searchable columns too
